### PR TITLE
openjdk: feature update to 17

### PR DESCRIPTION
**Pull request by liushuyu**
Topic Description
-----------------

openjdk: feature update to 17

Package(s) Affected
-------------------

```
openjdk
openjfx
```

Security Update?
----------------

No

Architectural Progress
----------------------

- [X] AMD64 `amd64`   
- [X] AArch64 `arm64`

Secondary Architectural Progress
--------------------------------

Architectural progress for "secondary," or experimental ports does not impede on merging of this topic.

- [x] PowerPC 64-bit (Little Endian) `ppc64el`
- [x] Loongson 3 `loongson3`

----

After the pull request is merged, all package(s) affected must be rebuilt against the `stable` Git tree and environment (only `stable` repository should be enabled in `sources.list`). This section marks the progress above.

Please, make sure the list of architectures below matches the ones above.

Post-Merge Architectural Progress
---------------------------------

- [ ] AMD64 `amd64`   
- [ ] AArch64 `arm64`

Post-Merge Secondary Architectural Progress
-------------------------------------------

Architectural progress for "secondary," or experimental ports does not impede on merging of this topic.

- [ ] Loongson 3 `loongson3`
- [ ] PowerPC 64-bit (Little Endian) `ppc64el`

### CODE REVIEW

**Comment by liushuyu** [comment]
TODO: OpenJDK 11 is still required and will be reinstated on this branch.

**Comment by liushuyu** [comment]
Ready for review.

The following items are still not done and they may as well be addressed in a later topic:

- Check for all incompatible reverse-dependencies (Currently, only some of them are checked and marked)
- Cherry-pick some patches from https://github.com/JetBrains/JetBrainsRuntime to JDK 11 to fix certain crashes

File: extra-java/openjdk-11/autobuild/beyond [hunk 1]
```diff
@@ -0,0 +1,11 @@
+printf 'update-alternatives --install /usr/bin/java java /usr/lib/java-11/bin/java 50' >> "$SRCDIR/autobuild/postinst"
```
Comment (MingcongBai) [inline extra-java/openjdk-11/autobuild/beyond:2]: abinfo missing throughout.

File: extra-java/openjdk-11/autobuild/build [hunk 2]
```diff
@@ -0,0 +1,94 @@
+##### Control part variables
+UPDATE=${PKGVER:2:-3}
```
Comment (MingcongBai) [inline extra-java/openjdk-11/autobuild/build:2]: Code style non-compliant.

File: extra-java/openjdk-11/autobuild/conffiles [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/etc/openjdk/ppc/jvm.cfg
```
Comment (MingcongBai) [inline extra-java/openjdk-11/autobuild/conffiles]: This probably could be generated.

File: extra-java/openjdk-11/autobuild/patches/debian-hotspot-mips-align.diff [hunk 4]
```diff
@@ -0,0 +1,13 @@
+Index: b/src/hotspot/cpu/zero/cppInterpreter_zero.cpp
```
Comment (MingcongBai) [inline extra-java/openjdk-11/autobuild/patches/debian-hotspot-mips-align.diff]: Patch file name non-compliant.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

File: extra-java/openjdk-11/autobuild/prepare [hunk 5]
```diff
@@ -0,0 +1,4 @@
+export EXTRA_CFLAGS="${CPPFLAGS} ${CFLAGS} -Wno-error"
```
Comment (MingcongBai) [inline extra-java/openjdk-11/autobuild/prepare:2]: abinfo missing throughout.

File: extra-java/openjdk/autobuild/build [hunk 6]
```diff
@@ -52,32 +35,35 @@ unset JAVA_HOME
 unset _JAVA_OPTIONS
 
 install -d -m 755 "$SRCDIR"/image
```
Comment (MingcongBai) [inline extra-java/openjdk/autobuild/build]: Verbosity here.

File: extra-java/openjdk/autobuild/build [hunk 7]
```diff
@@ -52,32 +35,35 @@ unset JAVA_HOME
...
+abinfo "Configuring OpenJDK ${PKGVER} ..."
 sh configure \
```
Comment (MingcongBai) [inline extra-java/openjdk/autobuild/build]: Absolute path here.

File: extra-java/openjdk/autobuild/build [hunk 8]
```diff
@@ -52,32 +35,35 @@ unset JAVA_HOME
...
+    --with-jvm-features="${EXTRA_FEATURES}" \
     --enable-precompiled-headers \
     --with-extra-cxxflags="${EXTRA_CPP_FLAGS}" \
     --with-extra-cflags="${EXTRA_CFLAGS}" \
     --with-extra-ldflags="${LDFLAGS}"
 
-# FIXME: use system giflibwhen possible
-
-make images docs
+make product-images docs
```
Comment (MingcongBai) [inline extra-java/openjdk/autobuild/build:57]: abinfo missing,

File: extra-java/openjdk/autobuild/build [hunk 9]
```diff
@@ -52,32 +35,35 @@ unset JAVA_HOME
...
 
 mkdir -p "$PKGDIR"/usr/lib/java
```
Comment (MingcongBai) [inline extra-java/openjdk/autobuild/build]: Verbosity here, and below.

File: extra-java/openjdk/autobuild/patches/debian-hotspot-mips-align.diff [hunk 10]
```diff
@@ -1,13 +1,11 @@
-Index: b/src/hotspot/cpu/zero/cppInterpreter_zero.cpp
```
Comment (MingcongBai) [inline extra-java/openjdk/autobuild/patches/debian-hotspot-mips-align.diff]: Patch file name non-compliant.

File: extra-java/openjdk/autobuild/prepare [hunk 11]
```diff
@@ -1,4 +1,4 @@
-export EXTRA_CFLAGS="${CPPFLAGS} ${CFLAGS} -Wno-error"
-export EXTRA_CPP_FLAGS="${CPPFLAGS} ${CXXFLAGS} -Wno-error"
+export EXTRA_CFLAGS="${CPPFLAGS} ${CFLAGS/-fexceptions/}"
```
Comment (MingcongBai) [inline extra-java/openjdk/autobuild/prepare:3]: abinfo missing throughout.

File: extra-java/openjdk/spec [hunk 12]
```diff
@@ -1,4 +1,3 @@
-VER=11.0.12+ga
-SRCS="tbl::https://openjdk-sources.osci.io/openjdk11/openjdk-${VER/+/-}.tar.xz"
-CHKSUMS="sha256::5f139e8760d1ea0587d029b3c217654ed3bf6f46d663eb418e19f42f36c061e5"
-CHKUPDATE="anitya::id=177935"
+VER=17+ga
+SRCS="tbl::https://github.com/openjdk/jdk17u/archive/refs/tags/jdk-${VER/+/-}.tar.gz"
+CHKSUMS="sha256::8971e4c3555fffae60bd58a5fb403e51a1e0dda14645c6214b81ceff954dd411"
```
Comment (MingcongBai) [inline extra-java/openjdk/spec]: Missing CHKUPDATE.

File: extra-java/openjdk/autobuild/build [hunk 13]
```diff
@@ -34,12 +13,16 @@ SJAVAC=0
 [ "${CROSS:-$ARCH}" = "ppc64el" ] && export JARCH="ppc64le" CONFARCH="ppc64le"
 
 if [[ "${CROSS:-$ARCH}" != "amd64" && "${CROSS:-$ARCH}" != armel && "${CROSS:-$ARCH}" != ppc64* && "${CROSS:-$ARCH}" != "arm64" ]]; then
-    config_zero="--with-jvm-variants=zero --enable-precompiled-headers"
```
Comment (MingcongBai) [inline extra-java/openjdk/autobuild/build:37]: abinfo missing.

File: extra-java/openjdk/autobuild/build [hunk 14]
```diff
@@ -34,12 +13,16 @@ SJAVAC=0
...
+    config_zero="--with-jvm-variants=zero"
     SJAVAC=0
 else
     config_zero=" "
 fi
 
+if [[ "${CROSS:-$ARCH}" = "amd64" || "${CROSS:-$ARCH}" = "arm64" ]]; then
+    EXTRA_FEATURES+=",shenandoahgc"
```
Comment (MingcongBai) [inline extra-java/openjdk/autobuild/build]: abinfo missing.

File: extra-java/openjdk-11/autobuild/beyond (hunk 1 above)
Comment (MingcongBai) [inline extra-java/openjdk-11/autobuild/beyond:2]: Also verbosity missing throughout.

File: extra-java/openjfx-11/autobuild/build [hunk 15]
```diff
@@ -0,0 +1,51 @@
+export CXXFLAGS="${CXXFLAGS} -Wno-expansion-to-defined -fPIC "
```
Comment (MingcongBai) [inline extra-java/openjfx-11/autobuild/build:2]: abinfo and verbosity missing throughout.

File: extra-java/openjdk-11/autobuild/conffiles (hunk 3 above)
Comment (liushuyu) [reply extra-java/openjdk-11/autobuild/conffiles:1]: This only applies to ppc

File: extra-java/openjfx-11/autobuild/gradle.properties [hunk 16]
```diff
@@ -0,0 +1,5 @@
+COMPILE_WEBKIT = true
+COMPILE_MEDIA = true
+BUILD_JAVADOC = true
+BUILD_SRC_ZIP = true
+
```
Comment (MingcongBai) [inline extra-java/openjfx-11/autobuild/gradle.properties:5]: Trailing blank line.

File: extra-java/openjfx-11/autobuild/patches/0002-remove-arch-restriction.patch [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
++++ b/build.gradle	2019-01-26 22:30:49.827445418 -0700
+@@ -307,7 +307,7 @@
+     fail("Unknown and unsupported build architecture: $OS_ARCH")
+ } else if (IS_MAC && OS_ARCH != "x86_64") {
+     fail("Unknown and unsupported build architecture: $OS_ARCH")
+-} else if (IS_LINUX && OS_ARCH != "i386" && OS_ARCH != "amd64") {
++} else if (IS_LINUX && false) {
+     fail("Unknown and unsupported build architecture: $OS_ARCH")
+ }
+
```
Comment (MingcongBai) [inline extra-java/openjfx-11/autobuild/patches/0002-remove-arch-restriction.patch:11]: Tailing blank lines?

File: extra-java/openjfx-11/autobuild/patches/0004-fix-deprecation-error.patch [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
++++ b/build.gradle	2019-11-03 12:30:49.827445418 -0700
+@@ -929,6 +929,7 @@
+                 headers = headerDir
+                 output(ccOutput)
+                 params.addAll(variantProperties.ccFlags)
++                params.add("-Wno-deprecated-declarations")
+                 compiler = variantProperties.compiler
+                 source(variantProperties.nativeSource)
+                 cleanTask.delete ccOutput
+
```
Comment (MingcongBai) [inline extra-java/openjfx-11/autobuild/patches/0004-fix-deprecation-error.patch:11]: Trailing blank line?

File: extra-java/openjfx-11/autobuild/prepare [hunk 19]
```diff
@@ -0,0 +1,19 @@
+abinfo 'Setting Java version to 11u'
+export JDK_HOME=/usr/lib/java-11/
+update-alternatives --set java /usr/lib/java-11/bin/java
+
+if [[ "${CROSS:-$ARCH}" = "amd64" ]]; then
```
Comment (MingcongBai) [inline extra-java/openjfx-11/autobuild/prepare]: Could just use !=.

File: extra-java/openjfx/spec [hunk 20]
```diff
@@ -1,4 +1,3 @@
-VER=11.0.12+4
-SRCS="tbl::https://github.com/openjdk/jfx11u/archive/refs/tags/${VER}.tar.gz"
-CHKSUMS="sha256::4efc71bf26993d37b6fdeb6c07146c68092447e79b0d90ca4e60b0d6aff4cc55"
-CHKUPDATE="anitya::id=177935"
+VER=17+18
+SRCS="tbl::https://github.com/openjdk/jfx/archive/refs/tags/$VER.tar.gz"
+CHKSUMS="sha256::624bb375419a1c9deabd550cc82fababe9cf0bd1d5f366290cb56310dc99ef54"
```
Comment (MingcongBai) [inline extra-java/openjfx/spec]: CHKUPDATE missing.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

File: extra-java/openjdk-11/autobuild/beyond (hunk 1 above)
Comment (MingcongBai) [reply extra-java/openjdk-11/autobuild/beyond:2]: Issue not addressed.

Comment (MingcongBai) [reply extra-java/openjdk-11/autobuild/beyond:2]: Issue not addressed.

File: extra-java/openjdk-11/autobuild/build (hunk 2 above)
Comment (MingcongBai) [reply extra-java/openjdk-11/autobuild/build:2]: abinfo still missing.

File: extra-java/openjdk-11/autobuild/prepare (hunk 5 above)
Comment (MingcongBai) [reply extra-java/openjdk-11/autobuild/prepare:2]: Issue not resolved.

File: extra-java/openjdk/autobuild/build (hunk 13 above)
Comment (MingcongBai) [reply extra-java/openjdk/autobuild/build:37]: Issue not resolved, please add abinfo also for branched routines.

File: extra-java/openjdk/autobuild/build (hunk 14 above)
Comment (MingcongBai) [reply extra-java/openjdk/autobuild/build:23]: Ditto above.

File: extra-java/openjfx-11/autobuild/build (hunk 15 above)
Comment (MingcongBai) [reply extra-java/openjfx-11/autobuild/build:1]: Issue not resolved.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

**Review by MingcongBai — APPROVED** [review]

**Pull request merged**